### PR TITLE
Fixes invalid T-SQL column comment for quoted table names.

### PR DESCRIPTION
**Pull request by metaturso**
|      Q           |   A
|------------- | -----------
| Type            | bug
| BC Break     | no
| Fixed issues | #4283 

#### Summary

This merge request fixes the issue whereby a quoted table or schema name was getting converted into a string literal, resulting in a T-SQL statement referencing an invalid table name like `N'[table_name]'`. 

The issues happened whenever the entity table name contained [Quoted Reserved Words](https://www.doctrine-project.org/projects/doctrine-orm/en/2.7/reference/basic-mapping.html#quoting-reserved-words), for example:

```php
/**
 * @ORM\Table(name="`the-model-table`")
 */
class TheModel
{
}
```

The issue is fixed within the `SQLServer2012Platform` base class by inserting the schema or table names as T-SQL identifiers when generating the call to `sp_addextendedproperty` for quoted names. 

In other words, because Doctrine converts a quoted name like `` `table-name` `` into a valid T-SQL identifier like `[table-name]` we needn't wrap the identifier into a string literal before passing it to `getAddExtendedPropertySQL`.

### CODE REVIEW

File: src/Platforms/SQLServer2012Platform.php [hunk 1]
```diff
@@ -363,19 +363,24 @@ public function getCreatePrimaryKeySQL(Index $index, $table): string
      * as column comments are stored in the same property there when
      * specifying a column's "Description" attribute.
      *
-     * @param string $tableName  The quoted table name to which the column belongs.
+     * @param string $tableName The quoted table name to which the column belongs.
      * @param string $columnName The quoted column name to create the comment for.
-     * @param string $comment    The column's comment.
```
Comment (metaturso) [inline src/Platforms/SQLServer2012Platform.php:368]: Untabified by accident. Should I undo this change?

Comment (metaturso) [reply src/Platforms/SQLServer2012Platform.php:368]: Never mind, the PHPCS caught this as an error. I'm going to undo this.

**Comment by greg0ire** [comment]
I think you should target 2.11.x

**Comment by metaturso** [comment]
I rebased my branch onto `2.11.x`. (`git rebase --onto 2.11.x master fix/tsql-comment-column`) and it's possible that this didn't play nice with AppVeyor's mergeability checks. 

I'm happy to open a fresh MR targeting the desired branch if you  prefer.

There's a few places left in `SQLServerPlatform` that need the same fix (e.g. callers of `get*ExtendedPropertySQL` methods.) I'll write tests and fix for those too.

How would you feel about a method which initialises both variables at once, e.g. `[$schemaSQL, $tableSQL] = $this->doodad($tableName)`?

**Comment by greg0ire** [comment]
> I'm happy to open a fresh MR targeting the desired branch if you prefer.

Please don't, it's great the way it is 👍

I will close and reopen, it might help with the checks.

File: tests/Functional/Ticket/DBAL4283Test.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+{
+    public function testAddingExtendedPropertyToUnquotedTableName(): void
+    {
+        $createTableSQL = $this->getCreateTableSQL('table_name');
+
+        self::assertEquals(
+            "EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', "
+            . "N'SCHEMA', 'dbo', N'TABLE', 'table_name', N'COLUMN', column_name",
+            $createTableSQL
+        );
```
Comment (greg0ire) [inline tests/Functional/Ticket/DBAL4283Test.php:35]:
```suggestion
        self::assertEquals(
            "EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', "
            . "N'SCHEMA', 'dbo', N'TABLE', 'table_name', N'COLUMN', column_name",
            $this->getCreateTableSQL('table_name')
        );
```

File: tests/Functional/Ticket/DBAL4283Test.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    public function testAddingExtendedPropertyToQuotedTableName(): void
+    {
+        $createTableSQL = $this->getCreateTableSQL('`quoted-table-name`');
+
+        self::assertEquals(
+            "EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', "
+            . "N'SCHEMA', 'dbo', N'TABLE', [quoted-table-name], N'COLUMN', column_name",
+            $createTableSQL
+        );
+    }
```
Comment (greg0ire) [inline tests/Functional/Ticket/DBAL4283Test.php:47]:
```suggestion
        self::assertEquals(
            "EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', "
            . "N'SCHEMA', 'dbo', N'TABLE', [quoted-table-name], N'COLUMN', column_name",
            $this->getCreateTableSQL('`quoted-table-name`')
        );
    }
```

File: tests/Functional/Ticket/DBAL4283Test.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function testAddingExtendedPropertyToQuotedTableInSchema(): void
+    {
+        $createTableSQL = $this->getCreateTableSQL('`custom.quoted-table-name`');
+
+        self::assertEquals(
+            "EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', "
+            . "N'SCHEMA', [custom], N'TABLE', [quoted-table-name], N'COLUMN', column_name",
+            $createTableSQL
+        );
```
Comment (greg0ire) [inline tests/Functional/Ticket/DBAL4283Test.php:57]:
```suggestion
        self::assertEquals(
            "EXEC sp_addextendedproperty N'MS_Description', N'(DC2Type:array)', "
            . "N'SCHEMA', [custom], N'TABLE', [quoted-table-name], N'COLUMN', column_name",
            $this->getCreateTableSQL('`custom.quoted-table-name`')
        );
```

File: tests/Functional/Ticket/DBAL4283Test.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    }
```
Comment (greg0ire) [inline tests/Functional/Ticket/DBAL4283Test.php:58]: Would it make sense to use a data provider instead? All 3 tests have the same logic, it would be great to be able to see just what changes in one place.

File: tests/Functional/Ticket/DBAL4283Test.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    public function getCreateTableSQL(string $quotedName): string
+    {
+        $driver   = $this->createMock(Driver::class);
```
Comment (greg0ire) [inline tests/Functional/Ticket/DBAL4283Test.php:62]: Since you perform no expectations, this is a stub:

```suggestion
        $driver   = $this->createStub(Driver::class);
```

**Comment by morozov** [comment]
This change requires a functional test that runs the updated SQL queries by a real database, not making assertions against the generated SQL and using mocks.

**Comment by metaturso** [comment]
Couldn't agree more!

The proposed unit test is good enough to show the problem exists not prove that it's fixed. In that sense, it helped expose the fact that the ALTER and DROP column/table cases are also bugged.

I won't get another chance to work on this probably until next weekend, when I plan to begin writing the new integration test against a real SQL Server instance before attempting a fix. I'll try to push the new version of the unit failing unit test sooner than that to arouse discussion about a clean fix.

In the meantime, could you let me know your requirements or preferences for this new test? For example, could you point me to an existing test that I can or should use as reference for level of abstraction, structure, assertion style, etc., just to get an idea?

On a related note, is the current Travis CI pipeline already configured to run SQL Server tests?

Thanks in advance for your time.

**Comment by morozov** [comment]
> In the meantime, could you let me know your requirements or preferences for this new test? For example, could you point me to an existing test that I can or should use as reference for level of abstraction, structure, assertion style, etc., just to get an idea?

As far as I understand the problem, you'll need to create a table with a name that has to be quoted. Look for the tests that call `dropAndCreateTable()` under [tests/Doctrine/Tests/DBAL/Functional](https://github.com/doctrine/dbal/tree/2.11.x/tests/Doctrine/Tests/DBAL/Functional).

> On a related note, is the current Travis CI pipeline already configured to run SQL Server tests?

Yes, see the [Travis](https://travis-ci.org/github/doctrine/dbal/builds/729378606) and [AppVeyor](https://ci.appveyor.com/project/doctrine/dbal/builds/35345923) builds.

**Comment by mvorisek** [comment]
I confirm this bug, @metaturso can you fix the PR so it can be merged quickly?

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php [hunk 7]
```diff
@@ -360,13 +360,18 @@ public function getCreatePrimaryKeySQL(Index $index, $table)
...
         if (strpos($tableName, '.') !== false) {
             [$schemaSQL, $tableSQL] = explode('.', $tableName);
-            $schemaSQL              = $this->quoteStringLiteral($schemaSQL);
-            $tableSQL               = $this->quoteStringLiteral($tableSQL);
-        } else {
-            $schemaSQL = "'dbo'";
-            $tableSQL  = $this->quoteStringLiteral($tableName);
+        }
+
+        if (strpos($tableName, '[') === false) {
```
Comment (mvorisek) [inline lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:369]: I think stripping the `[` `]` quotes would be better, as we can then quote it using `'` like we would normally do.

Comment (mvorisek) [inline lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:369]: I think we should `preg_replace('~^\[|\]$~s', '', $tableName)` instead and quote as literal as we would normally do.

String literal seeems to be the official way to escape: https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-addextendedproperty-transact-sql?view=sql-server-ver15#b-adding-an-extended-property-to-a-column-in-a-table

**Comment by mvorisek** [comment]
@metaturso please see https://github.com/doctrine/dbal/pull/4360 , I think we need to fix in a little more abstract way - please provide feedback, if you want to change anything or help with the tests, please send a PR directly to my branch. Michael

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php (hunk 7 above)
Comment (metaturso) [reply lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php:369]: > @metaturso please see #4360 , I think we need to fix in a little more abstract way - please provide feedback, if you want to change anything or help with the tests, please send a PR directly to my branch. Michael

@mvorisek thanks for stepping in with a patch for this, chap. Your fix in PR #4360 is reasonably self-contained and I'm happy to ditch my changes in favour of yours. 

There's definitely something about the amount of (un)escaping that's done at multiple levels that doesn't completely sit right with me, but fixing the data-flow is a much more pervasive change that's likely to break B/C or third-party code so I can't offer a better alternative.

I didn't get a chance to sit down and write the promised integration test. I'll commit and push the fixed version of the unit test, which runs against 2.11.x version of the codebase, during my lunch break. Feel free to steal this from my PR if that's of any use to you.

Should you decide not to use my test, please don't forget to cover for #4283 by adding test cases or fixture using i) table names with a custom schema name and ii) containing special characters or reserved keywords to your pull request.

**Comment by metaturso** [comment]
I've applied all suggestions to the unit test and updated it to cover the *add*, *update* and *drop* Extended Property cases, @morozov and @greg0ire thanks for your feedback and suggestions. 

All changes to the SQL Server platform class have been removed from this branch because @mvorisek has already raised a comprehensive fix to the original issue (thanks for that.) The failing unit test has been left here for his convenience.

I reckon this PR can now be closed in favour of #4360.

**Comment by mvorisek** [comment]
@metaturso please submit the tests as PR into https://github.com/mvorisek/dbal/tree/fix_mssql_comment_with_escaped_column branch, so your work can be integrated into #4360

**Comment by metaturso** [comment]
Will do shortly. Close this PR at your discretion.
